File: ProjectEulerPython/problems/problem_004.py
import int_helpers
from problems.problem import Problem

class Problem004(Problem):

  def calculate_answer(self) -> int:
    answer = 0

    self.print_detail("Incremental results:")
    for i in range(999, 99, -1):
      for j in range(999, i - 1, -1):
        product = i * j
        if product <= answer:
          break
        if int_helpers.is_palindrome(product):
          answer = product
          self.print_detail(f"{product} = {i} * {j}")
          break

    return answer

File: ProjectEuler/Problems/Problem004.cs
using Microsoft.Extensions.Logging;

namespace ProjectEuler.Problems
{
    public class Problem004 : Problem<int>
    {
        public Problem004(ILogger<Problem004> logger) : base(logger) { }

        protected override int CalculateAnswer()
        {
            var answer = 0;

            Logger.LogDebug("Incremental Results:");
            for (var i = 999; i >= 100; --i)
            {
                for (var j = 999; j >= i; --j)
                {
                    var product = i * j;
                    if (product <= answer) break;
                    if (product.IsPalindrome())
                    {
                        Logger.LogDebug("{product} = {i} * {j}", product, i, j);
                        answer = product;
                        break;
                    }
                }
            }

            return answer;
        }
    }
}
